### PR TITLE
plot elongation

### LINKED ISSUE #804: Cannot Plot Elongation
Seems to be a bug when trying to plot elongation:

```python
from desc.plotting import plot_1d
from desc.grid import LinearGrid
from desc.examples import get

eq=get("precise_QA")
plot_1d(eq,"a_major/a_minor",grid=LinearGrid(N=20,NFP=eq.NFP))
```

```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[36], line 1
----> 1 plot_1d(eq,"a_major/a_minor",grid=LinearGrid(N=20,NFP=eq.NFP))

File ~/DESC/desc/plotting.py:496, in plot_1d(eq, name, grid, log, ax, return_data, **kwargs)
    493 if len(plot_axes) != 1:
    494     return ValueError(colored("Grid must be 1D", "red"))
--> 496 data, ylabel = _compute(eq, name, grid, kwargs.pop("component", None))
    497 label = kwargs.pop("label", None)
    499 fig, ax = _format_ax(ax, figsize=kwargs.pop("figsize", None))

File ~/DESC/desc/plotting.py:314, in _compute(eq, name, grid, component, reshape)
    305 label = (
    306     r"$"
    307     + label
   (...)
    310     + ")$"
    311 )
    313 if reshape:
--> 314     data = data.reshape((grid.num_theta, grid.num_rho, grid.num_zeta), order="F")
    316 return data, label

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/numpy/array_methods.py:146, in _reshape(a, order, *args)
    144 elif order == "F":
    145   dims = list(range(a.ndim)[::-1])
--> 146   return lax.reshape(a, newshape[::-1], dims).T
    147 elif order == "A":
    148   raise NotImplementedError("np.reshape order=A is not implemented.")

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/lax/lax.py:862, in reshape(operand, new_sizes, dimensions)
    859 else:
    860   dyn_shape, static_new_sizes = _extract_tracers_dyn_shape(new_sizes)
--> 862   return reshape_p.bind(
    863     operand, *dyn_shape, new_sizes=tuple(static_new_sizes),
    864     dimensions=None if dims is None or same_dims else dims)

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/core.py:386, in Primitive.bind(self, *args, **params)
    383 def bind(self, *args, **params):
    384   assert (not config.jax_enable_checks or
    385           all(isinstance(arg, Tracer) or valid_jaxtype(arg) for arg in args)), args
--> 386   return self.bind_with_trace(find_top_trace(args), args, params)

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/core.py:389, in Primitive.bind_with_trace(self, trace, args, params)
    388 def bind_with_trace(self, trace, args, params):
--> 389   out = trace.process_primitive(self, map(trace.full_raise, args), params)
    390   return map(full_lower, out) if self.multiple_results else full_lower(out)

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/core.py:821, in EvalTrace.process_primitive(self, primitive, tracers, params)
    820 def process_primitive(self, primitive, tracers, params):
--> 821   return primitive.impl(*tracers, **params)

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/dispatch.py:131, in apply_primitive(prim, *args, **params)
    129 try:
    130   in_avals, in_shardings = util.unzip2([arg_spec(a) for a in args])
--> 131   compiled_fun = xla_primitive_callable(
    132       prim, in_avals, OrigShardings(in_shardings), **params)
    133 except pxla.DeviceAssignmentMismatchError as e:
    134   fails, = e.args

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/util.py:263, in cache.<locals>.wrap.<locals>.wrapper(*args, **kwargs)
    261   return f(*args, **kwargs)
    262 else:
--> 263   return cached(config._trace_context(), *args, **kwargs)

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/util.py:256, in cache.<locals>.wrap.<locals>.cached(_, *args, **kwargs)
    254 @functools.lru_cache(max_size)
    255 def cached(_, *args, **kwargs):
--> 256   return f(*args, **kwargs)

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/dispatch.py:222, in xla_primitive_callable(prim, in_avals, orig_in_shardings, **params)
    220     return out,
    221 donated_invars = (False,) * len(in_avals)
--> 222 compiled = _xla_callable_uncached(
    223     lu.wrap_init(prim_fun), prim.name, donated_invars, False, in_avals,
    224     orig_in_shardings)
    225 if not prim.multiple_results:
    226   return lambda *args, **kw: compiled(*args, **kw)[0]

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/dispatch.py:249, in _xla_callable_uncached(fun, name, donated_invars, keep_unused, in_avals, orig_in_shardings)
    247 def _xla_callable_uncached(fun: lu.WrappedFun, name, donated_invars,
    248                            keep_unused, in_avals, orig_in_shardings):
--> 249   computation = sharded_lowering(
    250       fun, name, donated_invars, keep_unused, True, in_avals, orig_in_shardings,
    251       lowering_platform=None)
    252   return computation.compile().unsafe_call

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/dispatch.py:241, in sharded_lowering(fun, name, donated_invars, keep_unused, inline, in_avals, in_shardings, lowering_platform)
    236 in_shardings = [UNSPECIFIED if i is None else i for i in in_shardings]  # type: ignore
    238 # Pass in a singleton `UNSPECIFIED` for out_shardings because we don't know
    239 # the number of output avals at this stage. lower_sharding_computation will
    240 # apply it to all out_avals.
--> 241 return pxla.lower_sharding_computation(
    242     fun, 'jit', name, in_shardings, UNSPECIFIED, donated_invars,
    243     tuple(in_avals), keep_unused=keep_unused, inline=inline, always_lower=False,
    244     devices_from_context=None, lowering_platform=lowering_platform)

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/profiler.py:314, in annotate_function.<locals>.wrapper(*args, **kwargs)
    311 @wraps(func)
    312 def wrapper(*args, **kwargs):
    313   with TraceAnnotation(name, **decorator_kwargs):
--> 314     return func(*args, **kwargs)
    315   return wrapper

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/interpreters/pxla.py:1899, in lower_sharding_computation(fun_or_jaxpr, api_name, fun_name, in_shardings, out_shardings, donated_invars, global_in_avals, keep_unused, inline, always_lower, devices_from_context, lowering_platform, override_lowering_rules)
   1893 # 1. Trace to jaxpr and preprocess/verify it
   1894 auto_spmd_lowering = (
   1895     check_if_any_auto(in_shardings) if is_unspecified(out_shardings) else
   1896     check_if_any_auto(it.chain.from_iterable([in_shardings, out_shardings])))  # type: ignore
   1898 (closed_jaxpr, global_in_avals, global_out_avals, donated_invars,
-> 1899  kept_var_idx, name_stack) = _trace_to_jaxpr_and_dce(
   1900     fun_or_jaxpr, global_in_avals, api_name, fun_name, keep_unused,
   1901     donated_invars, auto_spmd_lowering)
   1902 jaxpr = closed_jaxpr.jaxpr
   1903 in_shardings = tuple(s for i, s in enumerate(in_shardings) if i in kept_var_idx)

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/interpreters/pxla.py:1692, in cache_wrap.<locals>.wrapped(f, *args, **kwargs)
   1690 def wrapped(f, *args, **kwargs):
   1691   if isinstance(f, lu.WrappedFun):
-> 1692     return _wrapped_with_lu_cache(f, *args, **kwargs)
   1693   else:
   1694     return _wrapped_with_weakref_lru_cache(f, *args, **kwargs)

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/linear_util.py:345, in cache.<locals>.memoized_fun(fun, *args)
    343   fun.populate_stores(stores)
    344 else:
--> 345   ans = call(fun, *args)
    346   cache[key] = (ans, fun.stores)
    348 return ans

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/interpreters/pxla.py:1707, in _trace_to_jaxpr_and_dce(fun_or_jaxpr, global_in_avals, api_name, fun_name, keep_unused, donated_invars, auto_spmd_lowering)
   1703 if isinstance(fun_or_jaxpr, lu.WrappedFun):
   1704   with dispatch.log_elapsed_time(
   1705       "Finished tracing + transforming {fun_name} in {elapsed_time} sec",
   1706       fun_name=str(name_stack), event=dispatch.JAXPR_TRACE_EVENT):
-> 1707     jaxpr, global_out_avals, consts = pe.trace_to_jaxpr_final(
   1708         fun_or_jaxpr, global_in_avals)
   1709 else:
   1710   assert isinstance(fun_or_jaxpr, core.ClosedJaxpr)

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/profiler.py:314, in annotate_function.<locals>.wrapper(*args, **kwargs)
    311 @wraps(func)
    312 def wrapper(*args, **kwargs):
    313   with TraceAnnotation(name, **decorator_kwargs):
--> 314     return func(*args, **kwargs)
    315   return wrapper

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/interpreters/partial_eval.py:2233, in trace_to_jaxpr_final(fun, in_avals, debug_info, keep_inputs)
   2231   main.jaxpr_stack = ()  # type: ignore
   2232   with core.new_sublevel():
-> 2233     jaxpr, out_avals, consts = trace_to_subjaxpr_dynamic(
   2234       fun, main, in_avals, keep_inputs=keep_inputs, debug_info=debug_info)
   2235   del fun, main
   2236 return jaxpr, out_avals, consts

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/interpreters/partial_eval.py:2177, in trace_to_subjaxpr_dynamic(fun, main, in_avals, keep_inputs, debug_info)
   2175 in_tracers = _input_type_to_tracers(trace.new_arg, in_avals)
   2176 in_tracers_ = [t for t, keep in zip(in_tracers, keep_inputs) if keep]
-> 2177 ans = fun.call_wrapped(*in_tracers_)
   2178 out_tracers = map(trace.full_raise, ans)
   2179 jaxpr, consts = frame.to_jaxpr(out_tracers)

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/linear_util.py:188, in WrappedFun.call_wrapped(self, *args, **kwargs)
    185 gen = gen_static_args = out_store = None
    187 try:
--> 188   ans = self.f(*args, **dict(self.params, **kwargs))
    189 except:
    190   # Some transformations yield from inside context managers, so we have to
    191   # interrupt them before reraising the exception. Otherwise they will only
    192   # get garbage-collected at some later time, running their cleanup tasks
    193   # only after this exception is handled, which can corrupt the global
    194   # state.
    195   while stack:

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/dispatch.py:216, in xla_primitive_callable.<locals>.prim_fun(*args)
    215 def prim_fun(*args):
--> 216   out = prim.bind(*args, **params)
    217   if prim.multiple_results:
    218     return out

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/core.py:386, in Primitive.bind(self, *args, **params)
    383 def bind(self, *args, **params):
    384   assert (not config.jax_enable_checks or
    385           all(isinstance(arg, Tracer) or valid_jaxtype(arg) for arg in args)), args
--> 386   return self.bind_with_trace(find_top_trace(args), args, params)

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/core.py:389, in Primitive.bind_with_trace(self, trace, args, params)
    388 def bind_with_trace(self, trace, args, params):
--> 389   out = trace.process_primitive(self, map(trace.full_raise, args), params)
    390   return map(full_lower, out) if self.multiple_results else full_lower(out)

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/interpreters/partial_eval.py:1854, in DynamicJaxprTrace.process_primitive(self, primitive, tracers, params)
   1852 def process_primitive(self, primitive, tracers, params):
   1853   if primitive in custom_staging_rules:
-> 1854     return custom_staging_rules[primitive](self, *tracers, **params)
   1855   return self.default_process_primitive(primitive, tracers, params)

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/lax/lax.py:3362, in _reshape_staging_rule(trace, x, new_sizes, dimensions, *dyn)
   3360 params = dict(new_sizes=new_sizes, dimensions=dimensions)
   3361 if not dyn:
-> 3362   return trace.default_process_primitive(reshape_p, (x,), params)
   3363 av = core.DShapedArray(_merge_dyn_shape(new_sizes, dyn), x.dtype, x.weak_type)
   3364 return _dyn_shape_staging_rule(trace, reshape_p, av, x, *dyn, **params)

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/interpreters/partial_eval.py:1859, in DynamicJaxprTrace.default_process_primitive(self, primitive, tracers, params)
   1857 def default_process_primitive(self, primitive, tracers, params):
   1858   avals = [t.aval for t in tracers]
-> 1859   out_avals, effects = primitive.abstract_eval(*avals, **params)
   1860   # == serve as a "not xor" here.
   1861   if not (isinstance(out_avals, (tuple,list)) == primitive.multiple_results):

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/core.py:422, in _effect_free_abstract_eval.<locals>.abstract_eval_(*args, **kwargs)
    421 def abstract_eval_(*args, **kwargs):
--> 422   return abstract_eval(*args, **kwargs), no_effects

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/lax/utils.py:63, in standard_abstract_eval(prim, shape_rule, dtype_rule, weak_type_rule, named_shape_rule, *avals, **kwargs)
     61   return core.ConcreteArray(out.dtype, out, weak_type=weak_type)
     62 elif least_specialized is core.ShapedArray:
---> 63   return core.ShapedArray(shape_rule(*avals, **kwargs),
     64                           dtype_rule(*avals, **kwargs), weak_type=weak_type,
     65                           named_shape=named_shape_rule(*avals, **kwargs))
     66 elif least_specialized is core.DShapedArray:
     67   shape = shape_rule(*avals, **kwargs)

File ~/.conda/envs/desc-env/lib/python3.11/site-packages/jax/_src/lax/lax.py:3308, in _reshape_shape_rule(operand, new_sizes, dimensions)
   3305 if (not config.jax_dynamic_shapes and
   3306     not math.prod(np.shape(operand)) == math.prod(new_sizes)):
   3307   msg = 'reshape total size must be unchanged, got new_sizes {} for shape {}.'
-> 3308   raise TypeError(msg.format(new_sizes, np.shape(operand)))
   3309 if dimensions is not None:
   3310   if set(dimensions) != set(range(np.ndim(operand))):

TypeError: reshape total size must be unchanged, got new_sizes (41, 1, 1) for shape (18513,).
```

### PR DESCRIPTION
**Pull request by ddudt**
Resolves #804.  

Does the same grid resampling that we do for functions of $\rho$, but now also for $\zeta$. 

The `grid_type` data index tag is no longer used, but I keep it because we should start using this in the future. This PR resolves an issue about plotting functions of $\zeta$ vs $\rho$, but that tag will still be useful for determining the proper type of grid to use (e.g. `"quad"` vs `"linear"`). 

I added a new test and also changed the other 1D plotting tests to get examples so they don't depend on doing an equilibrium solve.

### CODE REVIEW

File: tests/test_plotting.py [hunk 1]
```diff
@@ -64,16 +64,28 @@ def test_kwarg_future_warning(DummyStellarator):
...
-    eq = load(load_from=str(SOLOVEV["desc_h5_path"]))[-1]
+    eq = get("SOLOVEV")
     fig, ax, data = plot_1d(eq, "p", figsize=(4, 4), return_data=True)
     assert "p" in data.keys()
     return fig
 
 
+@pytest.mark.unit
+@pytest.mark.mpl_image_compare(remove_text=True, tolerance=tol_1d)
+def test_1d_elongation():
```
Comment (f0uriest) [inline tests/test_plotting.py:78]: is this image missing from the commit?

**Review by f0uriest — DISMISSED** [review]

**Comment by ddudt** [comment]
@unalmis I need your help to make the computation of `"a_major/a_minor"` work on symmetric grids

**Comment by unalmis** [comment]
> @unalmis I need your help to make the computation of `"a_major/a_minor"` work on symmetric grids

I will work on this now

**Comment by unalmis** [comment]
@ddudt The computation is working correctly. Here is a jupyter notebook with a visual of what is discussed below [elongation_test.zip](https://github.com/PlasmaControl/DESC/files/13866106/elongation_test.zip). There I confirm the results we compute match the analytic result from sympy.

The reason we are getting a wrong value for `a_major/a_minor` in `test_elongation` on symmetric grids is a resolution issue. The two grids
```python
fg = LinearGrid(L=2, M=eq.M_grid, N=2, sym=False, NFP=eq.NFP)
sg = LinearGrid(L=2, M=eq.M_grid, N=2, sym=True, NFP=eq.NFP)
```
place nodes at different locations in $\theta$. To compute the elongation, we compute a perimeter of a constant $\zeta$ surface by integrating $\Vert \mathbf{e}_{\theta} \Vert$  over $\theta$.  For the particular surface used in `test_elongation`, the quadrature estimates the true integral better when we sample the integrand at the nodes used by `fg` than it does when we sample at the nodes used by `sg`. At reasonable resolution, this difference is not measurable because our midpoint scheme integration will converge to the true integral exponentially on the periodic domain over $\theta$. Increasing `eq.M_grid` from 2 to 4 will pass the test.

As a sanity check , after the fixes in #451 , we can construct a symmetric grid that has the same nodes as `fg` like this:
```python
unique_theta = fg.nodes[fg.unique_theta_idx, 1]
sg_manual = LinearGrid(
    L=2, theta=unique_theta[unique_theta <= np.pi], N=2, sym=True, NFP=eq.NFP
)
```
I have checked computations on `sg_manual` match `fg` at all resolutions.


(some details just for your information: The reason `sg` does not have the same node placement at `sg_manual` is just a design choice:
When a number is passed in for `L, M, N` (or rho, theta, zeta) on `LinearGrid`, we want to construct a grid where all the nodes are weighted equally in any integration etc. The simplest way to do this is to use node placement such that `grid.spacing` and `grid.weights` can be the the same across every node (like it is for `fg`).  When symmetry is set to `true`, it is only correct for `grid.spacing` and `grid.weights` to be the same across every node if the nodes are evenly spaced. This requires that the arc distance between the closest node to $\theta = \pi$ and its reflection across the $\theta = \pi$ axis (let's call this its left neighbor assuming counterclock wise is left) equals the distance to the node to its right. In particular, this means there can't be a node at $\theta = 0$ since that distance is 0, so `sg` must have a different node placement than `fg` in order for `sg.spacing` and `sg.weights` to be constant.

When the node placement is explicitly specified, as in `sg_manual`, it seems important to respect that placement and not change anything. For the grid `sg_manual`,  the attributes `sg.spacing` and  `sg.weights` are not constant - in particular, the nodes at $\theta \neq 0$ have double the weight and spacing as the node at $\theta = 0$ because we need to double count the ones at $\theta \neq 0$  since there would have been two of them if we had not deleted the nodes with $\theta > \pi$. This ensures that all computations on `sg_manual` match those of `fg`.
)


Let me know if there is another potential issue besides  `test_elongation`. As far as I'm aware, the only bug with grids in DESC is if the user gives wants to use grids with duplicate nodes to compute integrals. For example we prioritize correctness of surface integrals over line integrals as mentioned [in the docstring here](https://github.com/PlasmaControl/DESC/blob/b5921fd37016ef13999e5f1157d3ad0931602bc3/desc/compute/utils.py#L676). This could be fixed by creating an alias attribute `grid.line_spacing` that returns `grid.spacing` if there are no duplicate nodes or a different array of shape `(grid.num_nodes, 3)` when there are duplicates.

**Review by unalmis — APPROVED** [review]

**Comment by dpanici** [comment]
I have a question on how `master_compute_data.pkl` was updated, was the only failing key there the `elongation` and that is why it was updated? did you check that the other keys were still the same?

**Review by dpanici — CHANGES_REQUESTED** [review]
can you add to the `test_elongation` test function a case where you pass in the override grid, so that that part of the compute logic gets tested? Looks fine otherwise

**Comment by ddudt** [comment]
> I have a question on how `master_compute_data.pkl` was updated, was the only failing key there the `elongation` and that is why it was updated? did you check that the other keys were still the same?

Yes the only failing key was `elongation`, since the dimension was changed from a scalar to a vector.

**Comment by ddudt** [comment]
> can you add to the `test_elongation` test function a case where you pass in the override grid, so that that part of the compute logic gets tested? Looks fine otherwise

This is already covered by the `plot_1d` test of elongation. The patch coverage for this PR is 100% according to codecov

**Review by dpanici — APPROVED** [review]

**Pull request merged**